### ISSUE TITLE
Using torch.nn.functional.interpolate with 'nearest' mode introduces misalignment

### ISSUE THREAD

**Issue by Ali2500** [pytorch/pytorch#34808]
## 🐛 Bug

When resizing images and their corresponding segmentation masks, it is common practice to use bilinear interpolation for the images and nearest neighbor sampling for segmentation masks. However, I have observed that using the `interpolate` method in `torch.nn.functional` with `mode='nearest'` causes the resized segmentation masks to be shifted to the bottom right when compared to the original image.

## To Reproduce

This requires PIL and OpenCV to be installed. The code given loads an image and its corresponding segmentation mask and produces 4 visualizations of the mask overlayed on top of the image: 

1. Without any resizing, i.e., control baseline to assess mask accuracy.
2. Both image and mask are up-sampled using `torch.nn.functional.interpolate` with `mode='bilinear'` and `align_corners=False`.
3. The image is upsampled as in (2), but the mask is up-sampled with `mode='nearest'` (this is where the problem occurs).
4. The image is upsampled as in (2), but the mask is up-sampled using the `Image.resize` method in PIL. This appears to perform much better than the setting in (3).

The image and mask used can be downloaded from:
- Image: https://www.dropbox.com/s/gkgli16v080degr/image.jpg?dl=0
- Mask: https://www.dropbox.com/s/g4vgrdsdrzlg14s/mask.png?dl=0

Point the `IMAGE_PATH` and `MASK_PATH` variables in the code below to the location of these two files.

EDIT: Also added a visualization of the absolute difference between (3) and (4). This required matplotlib.

```
from PIL import Image

import cv2
import numpy as np
import torch
import torch.nn.functional as F

IMAGE_PATH = "/path/to/image.jpg"
MASK_PATH = "/path/to/mask.png"


def visualize_mask_on_image(image, mask):
    color = np.array([0, 255, 0], np.uint8)[None, None, :]
    masked_image = np.where(mask, color, image).astype(np.float32)
    masked_image = np.round((0.6 * masked_image) + (0.4 * image.astype(np.float32)))
    return masked_image.astype(np.uint8)


def resize_torch(interpolation_mode, align_corners):
    # open image and mask files
    image = cv2.imread(IMAGE_PATH, cv2.IMREAD_COLOR)
    mask = cv2.imread(MASK_PATH, cv2.IMREAD_UNCHANGED)

    # convert to torch tensor
    image = torch.from_numpy(image).permute(2, 0, 1).float()  # HWC to CHW and byte to float
    mask = torch.from_numpy(mask > 0).unsqueeze(0).float()

    # upsample image and mask by 4x
    image = F.interpolate(image.unsqueeze(0), size=(800, 1200), mode='bilinear', align_corners=False)
    mask = F.interpolate(mask.unsqueeze(0), size=(800, 1200), mode=interpolation_mode, align_corners=align_corners)

    # convert to numpy arrays
    image = image.squeeze(0).permute(1, 2, 0).byte().numpy()
    mask = (mask.squeeze(0).permute(1, 2, 0) > 0.5).byte().numpy()

    # overlay the mask on the image
    return visualize_mask_on_image(image, mask)


def resize_pil_nearest():
    # open image and mask files
    image = cv2.imread(IMAGE_PATH, cv2.IMREAD_COLOR)
    mask = Image.open(MASK_PATH)

    # convert segmentation mask to binary fg/bg mask
    colormap = [0] + [1 for _ in range(255)]
    mask = mask.point(colormap)

    # convert to torch tensor
    image = torch.from_numpy(image).permute(2, 0, 1).float()  # HWC to CHW and byte to float

    # upsample image and mask by 4x
    image = F.interpolate(image.unsqueeze(0), size=(800, 1200), mode='bilinear', align_corners=False)
    mask = np.array(mask.resize((1200, 800), resample=Image.NEAREST))

    # convert to numpy arrays
    image = image.squeeze(0).permute(1, 2, 0).byte().numpy()
    mask = mask[:, :, None]

    # overlay the mask on the image
    return visualize_mask_on_image(image, mask), mask


def no_resize_baseline():
    # open image and mask files
    image = cv2.imread(IMAGE_PATH, cv2.IMREAD_COLOR)
    mask = cv2.imread(MASK_PATH, cv2.IMREAD_UNCHANGED)[:, :, None]

    return visualize_mask_on_image(image, mask), mask


def plot_torch_pil_diff():
    _, mask_pil = resize_pil_nearest()
    _, mask_torch = resize_torch('nearest', None)

    mask_pil = mask_pil.astype(np.float32)
    mask_torch = mask_torch.astype(np.float32)

    diff = np.abs(mask_pil - mask_torch)[:, :, 0]
    diff_img = plt.imshow(diff)
    plt.colorbar(diff_img)
    plt.show()


resized_torch_bilinear, _ = resize_torch('bilinear', False)
resized_torch_nearest, _ = resize_torch('nearest', None)
resized_PIL_nearest, _ = resize_pil_nearest()
no_resize, _ = no_resize_baseline()

cv2.namedWindow('Torch Bilinear', cv2.WINDOW_NORMAL)
cv2.namedWindow('Torch Nearest', cv2.WINDOW_NORMAL)
cv2.namedWindow('PIL Nearest', cv2.WINDOW_NORMAL)
cv2.namedWindow('No Resize', cv2.WINDOW_NORMAL)

cv2.imshow('Torch Bilinear', resized_torch_bilinear)
cv2.imshow('Torch Nearest', resized_torch_nearest)
cv2.imshow('PIL Nearest', resized_PIL_nearest)
cv2.imshow('No Resize', no_resize)

plot_torch_pil_diff()
cv2.waitKey(0)
```

## Expected behavior

Settings (3) and (4) should result in roughly the same output, but they don't. Let's zoom into handle bars of the bicycles in the overlayed images to get a better idea:

Here's what setting (1) looks like (no resizing):

![No Resize_screenshot_16 03 2020](https://user-images.githubusercontent.com/14821941/76776998-d2ad6600-67a7-11ea-9045-33500c3ff924.png)

Here's what setting (2) looks like (both image and mask resized with bilinear):

![Torch Bilinear_screenshot_16 03 2020](https://user-images.githubusercontent.com/14821941/76776662-531f9700-67a7-11ea-9ec8-596a34ea15b2.png)

Here's what setting (3) looks like (mask resized with `mode='nearest'`). Notice the large black area close to the top right where the mask is not covering the object.

![Torch Nearest_screenshot_16 03 2020](https://user-images.githubusercontent.com/14821941/76776712-6599d080-67a7-11ea-94a0-b0cbea369132.png)

And finally here's what setting (4) looks like (mask resized using PIL's resize method with `Image.NEAREST` mode:

![PIL Nearest_screenshot_16 03 2020](https://user-images.githubusercontent.com/14821941/76776792-82ce9f00-67a7-11ea-99f1-9f1df83012b7.png)

Notice how (3) has a larger misalignment than (4). The fact that (2) is better is not surprising considering that it involves bilinear upsampling, but then I would expect PIL's resize method to perform similarly bad. As you can see though, (2) and (4) are actually similarly good whereas (3) is worse.

Just to confirm, let's look at the plot for the absolute difference between the resized masks produced by (3) and (4):

![diff](https://user-images.githubusercontent.com/14821941/76779539-cf1bde00-67ab-11ea-82f2-f9b1261f9cc8.png)

This might seem like a small difference, but it was wreaking havoc with my segmentation network training. My output masks would always be shifted to the right/bottom which was reducing the final IoU score by ~2%.

P.S. OpenCV's resize method seems to have the same problem as PyTorch's interpolate.

## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): Ubuntu 18.04.3 LTS
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): N/A
 - Python version: 3.7
 - CUDA/cuDNN version: 10.0.130/7.6.5
 - GPU models and configuration: GPU 0: GeForce GTX 980, GPU 1: GeForce GTX TITAN X
 - Any other relevant information: N/A

cc @albanD @mruberry @jbschlosser @fmassa @vfdev-5 @pmeier

**Comment by colesbury** [comment]
@fmassa can you take a look at this?

**Comment by wjgan7** [comment]
I think I've figured out both of their implementations. Seems like PyTorch doesn't do "true" nearest neighbor interpolation, probably for the sake of efficiency.

```
from PIL import Image

import numpy as np
import torch
import torch.nn.functional as F

MASK_PATH = "bike_mask.png"

mask = np.array(Image.open(MASK_PATH))

def resize_nearest_torch(mask):
    mask = torch.Tensor(mask).view(1, 1, 480, 854)
    mask = F.interpolate(mask, (800, 1200), mode='nearest')
    mask = mask.view(800, 1200).numpy()
    return mask

def resize_nearest_pil(mask):
    mask = Image.fromarray(mask)
    mask = mask.resize((1200, 800), resample=Image.NEAREST)
    mask = np.array(mask)
    return mask

def resize_torch_impl(mask):
    resized = np.zeros((800, 1200))
    for i in range(resized.shape[0]):
        for j in range(resized.shape[1]):
            x = 854 * j / 1200
            y = 480 * i / 800
            ii = int(y)
            jj = int(x)
            resized[i, j] = mask[ii, jj]
    return resized

def resize_pil_impl(mask):
    resized = np.zeros((800, 1200))
    for i in range(resized.shape[0]):
        for j in range(resized.shape[1]):
            x = 854 * j / 1200 + (854 / 1200 / 2) - 0.5
            y = 480 * i / 800 + (480 / 800 / 2) - 0.5
            ii = round(y)
            jj = round(x)
            resized[i, j] = mask[ii, jj]
    return resized

print(np.mean(np.abs(resize_nearest_torch(mask) - resize_nearest_pil(mask))))
print(np.mean(np.abs(resize_nearest_torch(mask) - resize_torch_impl(mask))))
print(np.mean(np.abs(resize_nearest_pil(mask) - resize_pil_impl(mask))))
```

**Comment by Ali2500** [comment]
I see. I think the PyTorch implementation should yield the same result as that obtained from PIL, especially since torchvision is integrated with PIL which gives the impression that their shared functionalities will yield similar results.

Or at the very least this should be documented somewhere and/or a new mode can be added to the `interpolate` method.

@fmassa

**Comment by wjgan7** [comment]
Yeah that makes sense. After looking into it a bit more, I think we would only need to change this line (230/231) in `nearest_neighbor_compute_source_index` in `aten/src/ATen/native/UpSample.h`.
```c
const int64_t src_index = std::min(static_cast<int64_t>(floorf(dst_index * scale)), input_size - 1);
```
to
```c
const int64_t src_index = std::min(static_cast<int64_t>(floorf((dst_index + 0.5) * scale)), input_size - 1);
```

**Comment by xvjiarui** [comment]
@colesbury 
Any update on this?

**Comment by Ali2500** [comment]
For now, I've resorted to always using bilinear interpolation for resizing masks. This is of course slower and more memory intensive since the mask tensor has to be converted to float, interpolated and then thresholded to give a binary mask.

@colesbury: Any plans on addressing this in a future release? I understand that this is a 'breaking change' in that changes the behavior of a function, but the current behavior is also quite problematic.

**Comment by fmassa** [comment]
@vfdev-5 I believe this might be related to the antialiasing in `nearest` mode that we were discussing, can you have a look?

**Comment by vfdev-5** [comment]
Following @wjgan7 comment https://github.com/pytorch/pytorch/issues/34808#issuecomment-638469752 
the difference between torch vs PIL is how output pixel is computed:
```
# Torch
            x = 854 * j / 1200
            y = 480 * i / 800
            ii = int(y)
            jj = int(x)
            resized[i, j] = mask[ii, jj]
# PIL
            x = 854 * (j + 0.5) / 1200 - 0.5
            y = 480 * (i + 0.5) / 800  - 0.5
            ii = round(y)
            jj = round(x)
            resized[i, j] = mask[ii, jj]
``` 

And this is exactly the implementation we have in PyTorch depending on `align_corners` or note (However, for nearest mode we have only `align_corners=true`):
https://github.com/pytorch/pytorch/blob/ee5a97de119cde561d9ac68fe15ea87fda46b059/aten/src/ATen/native/UpSample.h#L268-L271
and 
https://github.com/pytorch/pytorch/blob/ee5a97de119cde561d9ac68fe15ea87fda46b059/aten/src/ATen/native/cpu/UpSampleKernel.cpp#L582-L583

**Comment by ppwwyyxx** [comment]
`align_corners=True` shouldn't cause misalignment either. TensorFlow v1's nearest resize has align_corners=True option and that is a correct implementation. 

```python
import numpy as np
import tensorflow.compat.v1 as tf

arr = np.array([[0, 1, 2, 3, 4]], dtype='float32')

input = tf.constant(arr)
input4D = tf.reshape(input, [1, 1, 5, 1])
print(tf.image.resize(input4D, [1, 3], method='nearest', align_corners=True)[0,:,:,0])
print(tf.image.resize(input4D, [1, 2], method='nearest', align_corners=True)[0,:,:,0])
# tf.Tensor([[0. 2. 4.]], shape=(1, 3), dtype=float32)
# tf.Tensor([[0. 4.]], shape=(1, 2), dtype=float32)

import torch
import torch.nn.functional as F
arr = torch.from_numpy(arr[None, None, :, :])
print(F.interpolate(arr, (1, 3), mode='nearest')[0,0])
print(F.interpolate(arr, (1, 2), mode='nearest')[0,0])
# tensor([[0., 1., 3.]])
# tensor([[0., 2.]])

```

See my article for a detailed explanation: [Where Are Pixels](https://ppwwyyxx.com/blog/2021/Where-are-Pixels/)

**Comment by vfdev-5** [comment]
@ppwwyyxx thanks for the link, helps to understand the context ! 

So, according to it, scikit-image `transform.resize` and PIL both implement correctly the interpolation with nearest mode using (2) method. In case of `F.interpolate` method (2) can be effectively done using `align_corners=False` with nearest mode. Can we consider method (2) as a reference ?

> TensorFlow v1's nearest resize has align_corners=True option and that is a correct implementation.

If I perform the same resize with skimage and PIL they give a bit different result for the second case : 
```python
import numpy as np
from skimage.transform import resize
arr = np.array([[0, 1, 2, 3, 4]], dtype='float32')

print(resize(arr, (1, 3), order=0, preserve_range=True))
> [[0.02173508 2.         3.9782648 ]]
print(resize(arr, (1, 2), order=0, preserve_range=True))
> [[1.0311022 2.9688978]]


from PIL import Image

r = Image.fromarray(arr)
r = r.resize((3, 1), resample=Image.NEAREST)
r = np.array(r)
print(r)
> [[0. 2. 4.]]
r = Image.fromarray(arr)
r = r.resize((2, 1), resample=Image.NEAREST)
r = np.array(r)
print(r)
> [[1. 3.]]
```

**Comment by ppwwyyxx** [comment]
> Can we consider method (2) as a reference ?

Yes. If we're going to support one mode only, I believe this is the most common option: it's used in PIL, scikit-image, as well as TF2. (TF2 no longer has align_corners option). Here is the code & output using TF2:
```python
import numpy as np
import tensorflow as tf
arr = np.array([[0, 1, 2, 3, 4]], dtype='float32')
input = tf.constant(arr)
input4D = tf.reshape(input, [1, 1, 5, 1])
print(tf.image.resize(input4D, [1, 3], method='nearest')[0,:,:,0])  # [0, 2, 4]
print(tf.image.resize(input4D, [1, 2], method='nearest')[0,:,:,0])  # [1, 3]
```

**Comment by vfdev-5** [comment]
> If we're going to support one mode only

technically, user can not use `align_corners` with nearest mode due to 
https://github.com/pytorch/pytorch/blob/ee5a97de119cde561d9ac68fe15ea87fda46b059/torch/nn/functional.py#L3603-L3608

**Comment by jbschlosser** [comment]
Closing as addressed by addition of the new `nearest-exact` mode in #64501.
